### PR TITLE
Replace usage of ModuleRepository with interface

### PR DESCRIPTION
**Pull request by nenes25**
| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | develop
| Description?      |  For some internal needs i want to decorate the class \PrestaShop\PrestaShop\Core\Addon\Module\ModuleRepository .  <br />Unfortunately it is not possible with current code as some classes depends on the concrete class rather than on the interface \PrestaShop\PrestaShop\Core\Addon\Module\ModuleRepositoryInterface .<br />This PR will fix it
| Type?             | bug fix 
| Category?         | BO
| BC breaks?        |  yes
| Deprecations?     | no
| Fixed ticket?     | 
| How to test?      |  1. You can find a demo module in attachment which decorate the class \PrestaShop\PrestaShop\Core\Addon\Module\ModuleRepository <br />2. Install it<br />3. On current develop branch it should raise an exception once you try to login to the BO<br />4. With the fix it works as expected<br />5. ( If you want to test the module behavior you will need to create an admin restricted role with authorized access to modules management and to modules contactform, ps_checkpayment, hhdemoinstance. With this user you should only show this modules in the list )
| Possible impacts? |  If the affected classes call functions that are not in the interface the code will break.

**BC Breaks :**
* Added method `public function isActive(): bool` to the interface `\PrestaShop\PrestaShop\Core\Addon\Module\ModuleInterface `
* Modified method `public function getFilteredList(AddonListFilter $filter, bool $skip_main_class_attributes = false);` to `\PrestaShop\PrestaShop\Core\Addon\Module\ModuleRepositoryInterface`

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/27802)

### CODE REVIEW

**Comment by nenes25** [comment]
You can check the behavior with the module in attachment
[hhdemoinstance.zip](https://github.com/PrestaShop/PrestaShop/files/8148779/hhdemoinstance.zip)

**Comment by nenes25** [comment]
Well  static analysis shows that it will not work, i'll check if i can fix.

**Comment by saulaski** [comment]
> Well static analysis shows that it will not work, i'll check if i can fix.

`ModuleInterface` has no `isActive()` method, yet the method is implemented in `Module` class. It's not totally safe to fix the interface without any BC break to document in this PR.
Same about `ModuleRepositoryInterface` and `ModuleRepository`.

**Comment by nenes25** [comment]
@saulaski thanks for your feedback !
Do you think that i have to add a BC break and update the interfaces as well ?
My case seems to be relatively marginal but the best practices are to use interfaces as much as possible.

**Comment by saulaski** [comment]
@nenes25 
Totally agreed, injecting interfaces is the way to go in a proper OO design.

I suppose all modules extend the `Module` base class that already implements `ModuleInterface`. Then in theory updating the interface should be fine and cause no BC break.
Other reviewers might know better than I do, and confirm if it is safe or not.

**Review by tegbessou — DISMISSED** [review]
The code looks good for me!

**Review by saulaski — DISMISSED** [review]

**Comment by saulaski** [comment]
LGTM

**Review by matks — COMMENTED** [review]
I'm not sure an interface is the right solution here.

@tegbessou @saulaski I'm open to discussion. IMO we should add an interface when we expect multiple implementations to be used. I would say we should add an interface as soon as have a clear standard usecase where 2 implementations exist (example: a PaymentModule Interface because we know there will be multiple PaymentModules).

Adding interfaces to the software is impactful because interfaces are hard public contracts. Adding a method to an interface is a big BC break because all implementations are faulty if they don't add the missing method, so basically interfaces stay around unchanged for long.

@nenes25 would it be possible to achieve your need with a class that extends ModuleRepository or decorates it instead?

**Comment by saulaski** [comment]
@matks Then why the Module base class already implements an inconsistent interface ? It makes no sense. Same about ModuleRepository.
Correct me if I am wrong, but no custom module could work properly without the missing method `isActive()`. It means most modules have to extend the Module class, or implement the interface that doesn't include a mandatory method in public contract.
Interfaces are extremly important to use dependency injection.

**Comment by matks** [comment]
> @matks Then why the Module base class already implements an inconsistent interface ? It makes no sense. Same about ModuleRepository. Correct me if I am wrong, but no custom module could work properly without the missing method `isActive()`. It means most modules have to extend the Module class, or implement the interface that doesn't include a mandatory method in public contract. Interfaces are extremly important to use dependency injection.

Very good point 👍  I change my mind

File: src/Core/Addon/Module/ModuleInterface.php
Comment (matks) [inline src/Core/Addon/Module/ModuleInterface.php]:
```suggestion
     * Return true if module is active
```

**Review by matks — DISMISSED** [review]

File: src/Core/Addon/Module/ModuleInterface.php
Comment (Progi1984) [inline src/Core/Addon/Module/ModuleInterface.php]:
```suggestion
    public function isActive(): bool;
```

File: src/Core/Addon/Module/ModuleRepositoryInterface.php
Comment (Progi1984) [inline src/Core/Addon/Module/ModuleRepositoryInterface.php]:
```suggestion
    public function getModuleAttributes(string $name): ParameterBag;
```

**Review by Progi1984 — COMMENTED** [review]
Small feedback

File: src/Core/Addon/Module/ModuleInterface.php
Comment (nenes25) [reply src/Core/Addon/Module/ModuleInterface.php]: @Progi1984 i first already add this signature to the method but Phpstan fail as the return of src/Core/Addon/Module/Module:isActive is not signed.
Should i update this class too ?

File: src/Core/Addon/Module/ModuleRepositoryInterface.php
Comment (nenes25) [reply src/Core/Addon/Module/ModuleRepositoryInterface.php]: Same question as for isActive, should i update the class \PrestaShop\PrestaShop\Core\Addon\Module\ModuleRepository too ?

File: src/Core/Addon/Module/ModuleInterface.php
Comment (Progi1984) [reply src/Core/Addon/Module/ModuleInterface.php]: Yes. In all cases, you create a BC Break like you define new methods in interfaces. In this case, it's better to typehint methods.

**Comment by atomiix** [comment]
Hi @nenes25, there's a rework of the ModuleManager in progress, maybe have a look at it? https://github.com/PrestaShop/PrestaShop/pull/27632

File: src/Core/Addon/Module/ModuleInterface.php
Comment (saulaski) [reply src/Core/Addon/Module/ModuleInterface.php]: If until today all modules had to extend `Module` base class and can't implement `ModuleInterface` as an alternative implementation, then there is no BC break from the new method in the interface.
But typing the method will create a BC break.
Question: do all modules have to extend `Module` base class because it is the expected type instead of the interface ?

Comment (matks) [reply src/Core/Addon/Module/ModuleInterface.php]: > Question: do all modules have to extend Module base class because it is the expected type instead of the interface ?

Yes, as stated in the developer documentation => https://devdocs.prestashop.com/8/modules/creation/tutorial/

I guess this is because Modules not only need to provide some methods, there are also properties that have to be defined. The documentation mentions them.

File: src/Adapter/Module/Configuration/ModuleSelfConfigurator.php
Comment (Amoifr) [inline src/Adapter/Module/Configuration/ModuleSelfConfigurator.php:72]: we can also replace the annotation by a type int may be ?
```suggestion
    protected ModuleRepositoryInterface $moduleRepository;
```

File: src/PrestaShopBundle/EventListener/ModuleActivatedListener.php
Comment (Amoifr) [inline src/PrestaShopBundle/EventListener/ModuleActivatedListener.php:71]: same as previous comment. May be we can replace annotation by a typeInt. 
```suggestion
    private ModuleRepositoryInterface $moduleRepository;
```

File: src/PrestaShopBundle/Twig/HookExtension.php
Comment (Amoifr) [inline src/PrestaShopBundle/Twig/HookExtension.php:54]: same as previous annotation -> typeInt
```suggestion
    private ?ModuleRepositoryInterface $moduleRepository;
```

File: src/PrestaShopBundle/Twig/HookExtension.php
Comment (Amoifr) [inline src/PrestaShopBundle/Twig/HookExtension.php:65]:
```suggestion
        ?ModuleRepositoryInterface $moduleRepository = null
```

File: src/Core/Addon/Module/ModuleInterface.php
Comment (saulaski) [reply src/Core/Addon/Module/ModuleInterface.php]: Possibly yes... a part of my soul dies each time I meet a public prop :p

File: src/Adapter/Module/Configuration/ModuleSelfConfigurator.php
Comment (saulaski) [reply src/Adapter/Module/Configuration/ModuleSelfConfigurator.php:72]: Would be great indeed, but only from PHP 7.4 :(

File: src/PrestaShopBundle/Twig/HookExtension.php
Comment (saulaski) [reply src/PrestaShopBundle/Twig/HookExtension.php:65]: Actually `$moduleRepository` dependency is mandatory and shouldn't be optional. Take a look at line 138.
```suggestion
        ModuleRepositoryInterface $moduleRepository
```

**Review by saulaski — CHANGES_REQUESTED** [review]

File: src/Core/Addon/Module/ModuleRepositoryInterface.php
Comment (jolelievre) [reply src/Core/Addon/Module/ModuleRepositoryInterface.php]: I think it's ok to introduce a BC break for ModuleRepository since it's mainly an internal class/service in the core There are very few modules that should extend it and risk having mismatching prototypes

File: src/Core/Addon/Module/ModuleRepositoryInterface.php
Comment (jolelievre) [inline src/Core/Addon/Module/ModuleRepositoryInterface.php:67]: If you look at the `ModuleRepository` instance it seems the method has additional parameters:

`public function getModule($name, $skip_main_class_attributes = false, bool $cache = true)`

Maybe it's worth integrating them in the interface as well?

File: src/Core/Addon/Module/ModuleInterface.php
Comment (jolelievre) [reply src/Core/Addon/Module/ModuleInterface.php]: I don't know if you noticed but this interface is related to this object:
https://github.com/PrestaShop/PrestaShop/blob/cc853e6d0ce56c01648d0a530d929ac04297d9ff/src/Adapter/Module/Module.php

It's an internal abstraction used in the core, NOT the `Module` class which is used by all modules implementation So it's not that critical if we change it So we probably can add the strict type as well

File: src/Core/Addon/Module/ModuleRepositoryInterface.php
Comment (jolelievre) [reply src/Core/Addon/Module/ModuleRepositoryInterface.php:67]: Same for `getList` which also has another `$skip_main_class_attributes` (and it might be worth checking other methods as well)

File: src/PrestaShopBundle/Twig/HookExtension.php
Comment (jolelievre) [reply src/PrestaShopBundle/Twig/HookExtension.php:54]: We can't since we need to be compatible with PHP 7.1 for now

File: src/PrestaShopBundle/Twig/HookExtension.php
Comment (jolelievre) [reply src/PrestaShopBundle/Twig/HookExtension.php:65]: I agree it seems weird, and yet when you look at the service definition it is consistent:
https://github.com/PrestaShop/PrestaShop/blob/88bce71fb5def465a09d57bf22f20c62ede057a5/src/PrestaShopBundle/Resources/config/services/bundle/twig.yml#L53

**Review by jolelievre — COMMENTED** [review]
IMHO all these interfaces and classes are mostly used internally (the confusion with the legacy `\ModuleCore` class is easy but it's not the same class!) So it's ok to introduce the BC breaks and even to improve the methods with strict types and maybe update the missing parameters

File: src/PrestaShopBundle/Twig/HookExtension.php
Comment (saulaski) [reply src/PrestaShopBundle/Twig/HookExtension.php:65]: Perhaps because if there is no employee in context, then there is no HTML to render ?
On the other hand, the argument never allowed null until now. Looks like pure tricky luck :))

File: src/Core/Addon/Module/ModuleRepositoryInterface.php
Comment (nenes25) [reply src/Core/Addon/Module/ModuleRepositoryInterface.php:67]: Hi @jolelievre 
Thanks for your feedback, i've updated the function getFilteredList in the interface to match with current \PrestaShop\PrestaShop\Core\Addon\Module\ModuleRepository::getFilteredList signature.
But for the function getModule and getList the signature seems to be the good ones.

**Comment by nenes25** [comment]
Thanks all for your feedbacks :smile: 
I've updated this PR with all of them that can be applied.
Let me know if additional changes are required.

File: src/Core/Addon/Module/ModuleRepository.php
Comment (saulaski) [inline src/Core/Addon/Module/ModuleRepository.php]: phpdoc comments are already defined in ModuleRepositoryInterface, no need to double the workload.
Best solution is to provide no phpdoc at all in implementation classes, because most IDEs can analyze the interface signature from the implementation.
Or you could refer to interface through documentation inheritance:
```suggestion
    /**
     * {@inheritDoc}
     */
```

**Review by saulaski — CHANGES_REQUESTED** [review]

**Comment by saulaski** [comment]
@nenes25 CI has been fixed. You should rebase your commits and force push to restart CI tests.

**Review by saulaski — DISMISSED** [review]

File: src/Core/Addon/Module/ModuleRepositoryInterface.php
Comment (Progi1984) [inline src/Core/Addon/Module/ModuleRepositoryInterface.php]:
```suggestion
    public function getFilteredList(AddonListFilter $filter, bool $skip_main_class_attributes = false);
```

**Review by Progi1984 — COMMENTED** [review]
@nenes25 A last comment and it's good.

**Review by saulaski — APPROVED** [review]

**Review by Progi1984 — APPROVED** [review]

**Review by matks — APPROVED** [review]

**Review by PierreRambaud — APPROVED** [review]

**Comment by atomiix** [comment]
If we could wait for #27632 to be merged before, it would be great! :D

**Review by khouloudbelguith — COMMENTED** [review]
Hi @nenes25,

Could you please check the Conflicting files?

Thanks!

**Comment by nenes25** [comment]
Hello,

Well with the rework of the module management, almost everything that have been done here become useless.
I need to restart from my initial need and to see how it can be fulfilled now.

Regards,

**Comment by matks** [comment]
Following @nenes25 last comment and merge of https://github.com/PrestaShop/PrestaShop/pull/27632 I think this might not be useful anymore

I will follow [our guidelines](https://devdocs.prestashop.com/1.7/project/maintainers-guide/reviewing-pull-requests/#stale-pull-requests) and I'll close this pull request as it seems stale. Feel free to reopen or open another one if you think it's still relevant.

Thanks!

**Pull request closed without merging**
